controller/network/network.go

```diff
@@ -321,35 +341,44 @@ func (network *Network) CreateSession(srcR *Router, clientId *identity.TokenId,
 		}
 
 		// 4a: Create Route Messages
-		rms, err := circuit.CreateRouteMessages(sessionId, terminator.GetAddress())
+		rms, err := circuit.CreateRouteMessages(attempt, sessionId, terminator.GetAddress())
 		if err != nil {
 			return nil, err
 		}
-
-		// 5: Route Egress
 		rms[len(rms)-1].Egress.PeerData = clientId.Data
-		peerData, err := sendRoute(circuit.Path[len(circuit.Path)-1], rms[len(rms)-1], network.options.TerminationTimeout)
+
+		// 5: Routing
+		logrus.Debugf("route attempt [#%d] for [s/%s]", attempt + 1, sessionId.Token)
+		peerData, cleanups, err := rs.route(attempt, circuit, rms, strategy, terminator)
+		for k, v := range cleanups {
+			allCleanups[k] = v
+		}
 		if err != nil {
-			strategy.NotifyEvent(xt.NewDialFailedEvent(terminator))
-			retryCount++
-			if retryCount > 3 {
-				return nil, err
-			} else {
+			logrus.Warnf("route attempt [#%d] for [s/%s] failed (%v)", attempt + 1, sessionId.Token, err)
+			attempt++
+			if attempt < uint32(network.options.CreateSessionRetries) {
 				continue
-			}
-		} else {
-			strategy.NotifyEvent(xt.NewDialSucceeded(terminator))
-		}
 
-		// 6: Create Intermediate Routes
-		for i := 0; i < len(circuit.Path)-1; i++ {
-			_, err = sendRoute(circuit.Path[i], rms[i], network.options.RouteTimeout)
-			if err != nil {
-				return nil, err
+			} else {
+				// revert successful routes
```

Comment:
not clear on why we cleanup if this is the last try, but not if it isn't

Reply:
Because the cleanup table is amended with the routers from each attempt. So the `allCleanups` map contains ALL the routers that would need to be cleaned up.

Reply:
And, cleaning up in intermediate attempts made for some strange races during testing.